### PR TITLE
[POAE7-2675] primitive type if-then/case-when/coalesce expr support i…

### PR DESCRIPTION
**Pull request by haiweizh**
…n next codegen

### What changes were proposed in this pull request?

[POAE7-2618] case/coalesce/if expr support in next codegen.
This story will be divided into four subtasks:
[POAE7-2675] primitive type if-then expr support in next codegen.
[POAE7-2781] primitive type multiple case-when expr support in next codegen.
[POAE7-2782] string type if/case expr support in next codegen.
[POAE7-2783] coalesce expr support in next codegen.
Implement [POAE7-2675]/[POAE7-2781]/[POAE7-2783] in this pr.

### Why are the changes needed?

[POAE7-2675] primitive type if-then expr support in next codegen.
[POAE7-2781] primitive type multiple case-when expr support in next codegen.
[POAE7-2783] coalesce expr support in next codegen.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

UTs.

### Which label does this PR belong to?

cider

### CODE REVIEW

File: src/cider/type/plan/CaseExpr.cpp [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+  const auto& expr_pair_list = get_expr_pair_list();
+  const auto& else_expr = get_else_ref();
+  CHECK_GT(expr_pair_list.size(), 0);
+  const auto case_ti = get_type_info();
+  if (case_ti.is_integer() || case_ti.is_time() || case_ti.is_decimal() ||
+      case_ti.is_fp() || case_ti.is_boolean()) {
+    const auto type =
+        case_ti.is_decimal() ? decimal_to_int_type(case_ti) : case_ti.get_type();
+    JITValuePointer value = func.createVariable(getJITTag(type), "case_when_value_init");
+    *value = func.createLiteral(getJITTag(type), 0);
```
Comment (yma11) [inline src/cider/type/plan/CaseExpr.cpp:227]: L226 and L227 should can be merged as `JITValuePointer value = func.createVariable(getJITTag(type), "case_when_value_init", 0);`?

Comment (haiweizh) [reply src/cider/type/plan/CaseExpr.cpp]: Done.

Comment (yma11) [reply src/cider/type/plan/CaseExpr.cpp]: Seems these lines can't be changed. Try reverting back and fix code style/lint issue at the same time.

File: src/cider/type/plan/CaseExpr.cpp [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+        case_ti.is_decimal() ? decimal_to_int_type(case_ti) : case_ti.get_type();
+    JITValuePointer value = func.createVariable(getJITTag(type), "case_when_value_init", 0);
+    JITValuePointer null = func.createVariable(JITTypeTag::BOOL, "case_when_null_init", false);
+    JITValuePointer is_case = func.createVariable(JITTypeTag::BOOL, "is_case_init", false);
+    for (const auto& expr_pair : expr_pair_list) {
+      func.createIfBuilder()
+          ->condition([&]() {
+            cider::exec::nextgen::utils::FixSizeJITExprValue cond(
+                expr_pair.first->codegen(context));
+            auto condition = !cond.getNull() && cond.getValue();
```
Comment (kd97) [inline src/cider/type/plan/CaseExpr.cpp:234]: need to judge cond.getValue()?

File: src/cider/type/plan/CaseExpr.cpp [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+            return condition;
+          })
+          ->ifTrue([&]() {
+            cider::exec::nextgen::utils::FixSizeJITExprValue then_jit_expr_value(
+                expr_pair.second->codegen(context));
+            value.replace(then_jit_expr_value.getValue());
+            null.replace(then_jit_expr_value.getNull());
+            *is_case = func.createLiteral(JITTypeTag::BOOL, true);
+          })
```
[the remark this answers is not in the document]
Comment (haiweizh) [reply src/cider/type/plan/CaseExpr.cpp:246]: Done.

File: src/cider/type/plan/CaseExpr.cpp (hunk 1 above)
Comment (haiweizh) [reply src/cider/type/plan/CaseExpr.cpp]: Done.

File: src/cider/type/plan/CaseExpr.cpp [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+    const auto type =
+        case_ti.is_decimal() ? decimal_to_int_type(case_ti) : case_ti.get_type();
+    JITValuePointer value =
+        func.createVariable(getJITTag(type), "case_when_value_init", 0);
+    JITValuePointer null =
+        func.createVariable(JITTypeTag::BOOL, "case_when_null_init", false);
+    JITValuePointer is_case =
+        func.createVariable(JITTypeTag::BOOL, "is_case_init", false);
+    for (const auto& expr_pair : expr_pair_list) {
+      func.createIfBuilder()
```
Comment (yma11) [inline src/cider/type/plan/CaseExpr.cpp:233]: For this logic, will it also work for sql like `case col_a IS NULL then col_b else col_c`?

Comment (haiweizh) [reply src/cider/type/plan/CaseExpr.cpp:233]: Yes.

Comment (haiweizh) [reply src/cider/type/plan/CaseExpr.cpp:233]: There is a case "SELECT col_int, CASE WHEN col_str IS NULL THEN 10 WHEN col_int > 3 THEN 20 ELSE 0 END FROM test" to verify.

**Review by yma11 — APPROVED** [review]

**Review by kd97 — APPROVED** [review]

**Pull request merged**